### DISCUSSION TITLE
Cached problem fact not being created for tests?

### DISCUSSION THREAD

**Discussion by hotungkhanh** [TimefoldAI/timefold-solver#1109] [General]
Hi,

In our program, we're using a Cached problem fact:

```
    @ProblemFactCollectionProperty
    public List<ConflictingUnit> calculateSoftUnitConflictList() {
        ArrayList<ConflictingUnit> out = new ArrayList<ConflictingUnit>();
        for (var first : units) {
            for (var second : units) {
                if (first.getUnitID() >= second.getUnitID()) {
                    continue;
                }
                int numStudents = 0;
                for (Student firstStudent : first.getStudents()) {
                    if (second.getStudents().contains(firstStudent)) {
                        numStudents++;
                    }
                }
                if (numStudents > 0) {
                    out.add(new ConflictingUnit(first, second, numStudents));
                }
            }
        }
        return out;
    }
```
```
private Constraint studentConflict(ConstraintFactory constraintFactory) {
        // A student can be in at most one unit at the same time.
        return constraintFactory
                // Select each pair of conflicting units.
                .forEach(ConflictingUnit.class)
                // Find the first unit.
                .join(Unit.class, equal(ConflictingUnit::getUnit1, Function.identity()))
                // Find the second unit.
                .join(Unit.class, equal((conflictingUnit, unit1) -> conflictingUnit.getUnit2(), Function.identity()),
                        // Check if the 2 units are on the same weekday ...
                        equal((conflictingUnit, unit1) -> unit1.getDayOfWeek(), Unit::getDayOfWeek),
                        // ... in the same timeslot ...
                        overlapping((conflictingUnit, unit1) -> unit1.getStartTime(),
                                (conflictingUnit, unit1) -> unit1.getEnd(),
                                Unit::getStartTime, Unit::getEnd))
                // ... and penalize each pair with a hard weight.
                .penalize(HardSoftScore.ofHard(1),
                        (conflictingUnit, unit1, unit2) -> conflictingUnit.getNumStudent())
                .asConstraint("Student conflict");

    }
```

However, the problem arises when we try to write a unit test to test a constraint.

```
@Test
void studentConflictTest() {
    Unit unit1 = new Unit(1, "unit1", DAY_OF_WEEK, START_TIME, DURATION, List.of(STUDENT1),true, ROOM1);
    Unit unit2 = new Unit(2, "unit2", DAY_OF_WEEK, START_TIME, DURATION, List.of(STUDENT1),false, ROOM2);
    Unit unit3 = new Unit(3, "unit3", DAY_OF_WEEK, START_TIME, DURATION, List.of(STUDENT2),true, ROOM3);

    constraintVerifier.verifyThat(TimetableConstraintProvider::studentConflict)
            .given(unit1, unit2, unit3)
            .penalizesBy(1);
}
```

Is it the case that Cached problem facts are not created for unit tests?

Please let me know.

Thank you!

**Comment by triceo** [comment]
Hello @hotungkhanh!

The test in question doesn't use the planning solution, and therefore cannot read the problem facts. If you want your problem facts to be taken into account, you'll have to add them yourself, just like you add `unit1, unit2, unit3`.

**Reply by hotungkhanh** [reply]
Hi @triceo, thank you for your prompt response!

We have tried what you suggested.

However, the problem is we have only the Unit class as @PlanningEntityCollectionProperty in Timetable, and so only Unit can be provided to the .given() method of constraintVerifier.

Therefore, we cannot add the ConflictingUnit to the test.

Do you know how to solve this?

Thank you so much!

**Reply by triceo** [reply]
The planning solution or the annotations are irrelevant to constraint verifier. They are not being used.

    @Test
    void studentConflictTest() {
        Unit unit1 = new Unit(1, "unit1", DAY_OF_WEEK, START_TIME, DURATION, List.of(STUDENT1),true, ROOM1);
        Unit unit2 = new Unit(2, "unit2", DAY_OF_WEEK, START_TIME, DURATION, List.of(STUDENT1),false, ROOM2);
        Unit unit3 = new Unit(3, "unit3", DAY_OF_WEEK, START_TIME, DURATION, List.of(STUDENT2),true, ROOM3);
        ConflictingUnit conflictingUnit = new ConflictingUnit(...);

        constraintVerifier.verifyThat(TimetableConstraintProvider::studentConflict)
            .given(unit1, unit2, unit3, conflictingUnit)
            .penalizesBy(1);
}

**Reply by hotungkhanh** [reply]
Oh, I see.

Thank you so much for your help! 😊
